controllers/namespacelabel/namespacelabel_controller.go

```diff
@@ -0,0 +1,317 @@
+	Production struct {
+		CreateStrategy struct {
+			CidrBlock string `json:"CidrBlock"`
+		} `json:"createStrategy"`
+	} `json:"production"`
+}
+
+var (
+	// Map that associates the labels in the namespace to actions to perform
+	// - Keys are labels that might be set to the namespace object
+	// - Values are functions that receive the value of the label, the reconcile request,
+	//   and the reconciler instance.
+	namespaceLabelBasedActions = map[string]func(string, ctrl.Request, *NamespaceLabelReconciler) error{
+		// Uninstall RHMI
+		"api.openshift.com/addon-rhmi-operator-delete": Uninstall,
+		// Uninstall MAO
+		"api.openshift.com/addon-managed-api-service-delete": Uninstall,
+		// Update CIDR value
+		"cidr": CheckCidrValueAndUpdate,
+	}
+
+	// Map that associates the labels on the configMap to actions to perform
+	// - Keys are labels that might be set to the configMap object
+	// - Values are functions that receive the value of the label, the reconcile request,
+	//   and the reconciler instance.
+	configMapLabelBasedActions = map[string]func(string, reconcile.Request, *NamespaceLabelReconciler) error{
+		// Uninstall RHMI
+		"api.openshift.com/addon-rhmi-operator-delete": Uninstall,
+		// Uninstall MAO
+		"api.openshift.com/addon-managed-api-service-delete": Uninstall,
+	}
+)
+
+func (r *NamespaceLabelReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Namespace{}).
+		For(&corev1.ConfigMap{}).
+		Complete(r)
+}
+
+func New(mgr manager.Manager) *NamespaceLabelReconciler {
+	watchNS, err := resources.GetWatchNamespace()
+	if err != nil {
+		panic("could not get watch namespace from namespacelabel controller")
+	}
+	namespaceSegments := strings.Split(watchNS, "-")
+	namespacePrefix := strings.Join(namespaceSegments[0:2], "-") + "-"
+	operatorNs := namespacePrefix + "operator"
+
+	return &NamespaceLabelReconciler{
+		Client:            mgr.GetClient(),
+		Scheme:            mgr.GetScheme(),
+		operatorNamespace: operatorNs,
+		log:               log,
+	}
+}
+
+// NamespaceLabelReconciler reconciles a namespace label object
+type NamespaceLabelReconciler struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	k8sclient.Client
+	Scheme *runtime.Scheme
+
+	operatorNamespace string
+	controller        controller.Controller
+	log               l.Logger
+}
+
+// +kubebuilder:rbac:groups="",resources=namespaces,verbs=list;get;watch;update
+
+// Reconcile : The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *NamespaceLabelReconciler) Reconcile(request ctrl.Request) (ctrl.Result, error) {
+	_ = context.Background()
+	// log := r.Log.WithValues("namespacelabel", request.NamespacedName)
+
+	if request.NamespacedName.Name == r.operatorNamespace {
```

Comment:
```suggestion
	if request.Namespace == r.operatorNamespace {
```
Probably due to the operator-sdk update, but `request.NamespacedName.Name` doesn't look to be namespace name anymore but is the name of the watched config map name (can be checked by debugging with a breakpoint at this line). 

Without the above change (or to `request.NamespacedName.Namespace`), this block of code won't be run anymore so probably would affect addon deletion uninstall flow :thinking:

Reply:
Interesting stuff. I think that this is unrelated to the operator-sdk but a bug that was introduced when adding the ConfigMap watch to this controller. The `ctrl.Request` struct embeds `types.NamespacedName` which means that the `.NamespacedName` is actually redundant:
```go
type Request struct {
	// NamespacedName is the name and namespace of the object to reconcile.
	types.NamespacedName
}
```

`request.Namespace` would work for the ConfigMap events but I'm not sure if it'd work with Namespace resources as they're cluster scoped. I think ideally we'd have 2 separate controllers, one for the namespace that would use `request.Name` and one for the ConfigMap that could watch specifically the ConfigMap we're looking for, and the common logic could be factored in a separate package.

A quicker fix would be to simply check for both, but it sounds pretty ugly to me:
```go
if request.Name == r.operatorNamespace || request.Namespace == r.operatorNamespace {
```

Reply:
@KevFan If it sounds good to you, I'll keep it as it is in this PR and create a separate ticket to improve the controller. I think the reason it still works is because the controller is reconciling every minute, so changes in the ConfigMap will be spotted eventually

Reply:
@sergioifg94 Sure, sounds good to me once it's flagged :+1: